# Detecting Fake News with Python

To build a model to accurately classify a piece of news as REAL or FAKE.

In [2]:
#importing necessary libararies to build the model.
import pandas as pd
import numpy as np
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [3]:
#loading the datasets to the dataframe.
df=pd.read_csv('news.csv')

In [4]:
#displaying the number of rows and columns in the dataframe.
df.shape

(6335, 4)

In [5]:
#Showing the top 5 entries in the dataset.
df.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [6]:
#get the labels from the dataframe.
labels=df.label
labels.head()

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [7]:
#split the dataset.
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)

In [8]:
#Initialize a tfidfvectorizer.
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)

In [9]:
#fit and transform train set, test set.
tfidf_train=tfidf_vectorizer.fit_transform(x_train) 
tfidf_test=tfidf_vectorizer.transform(x_test)

In [10]:
#Initialize a PassiveAgressiveClassifier
pac=PassiveAggressiveClassifier(max_iter=50)
pac.fit(tfidf_train,y_train)

PassiveAggressiveClassifier(max_iter=50)

In [11]:
#predict the test set and calculate accuracy score.
y_pred=pac.predict(tfidf_test)
score=accuracy_score(y_test,y_pred)
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 93.13%


In [13]:
#Build Confusion matrix
confusion_matrix(y_test,y_pred, labels=['FAKE','REAL'])

array([[592,  46],
       [ 41, 588]], dtype=int64)

So with this model, we have 592 true positives,588 true negatives,41 false positives, 49 false negatives.

# Summary

Today, we learned to detect fake news with Python. 
We took a political dataset, implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. 
We ended up obtaining an accuracy of 93.13% in magnitude.